### PR TITLE
5 response building define response elements

### PR DESCRIPTION
**Pull request by gwolf-011235**
#### Overview

Adds basic functionality to create HTTP response based on a HTTP request. 
Right now only base functionality is supported
- GET method
- Default Error pages
- Single string index
- MIME types are hardcoded

#### 1 ResponseBuilder

This is the main class of this PR. It is designed to be instantiated once and then reused for every response. 
- It receives the ConfigFile. The active Server is set via Setter (how the active server is chosen is not implemented yet)
- With buildResponse() a HTTP request will be translated to a response, which can be retrieved with getResponse()
- Response building is passed to subclasses TargetResourceHandler and ResponseBodyHandler. Last step is adding Statusline and Headers and body.
- Class has a Stringstream which is reused by reset() before building the next response.

#### 2 TargetResourceHandler 

The responsibility of this class is to locate the target resource. 
- Receives a vector of locations which are searched for the longest matching path
- The located file is then checked for its type (regular, directory, not existing) and actions are taken
- A HTTP response is passed as ref, to set targetResource, status, autindex and location

#### 3 ResponseBodyHandler
 
The responsibility of this class is to create the body depending on the located target resource.
- if HTTP response has error constructs error page
- if autoindex is true constructs autoindex with class AutoindexHandler
- if method is GET reads the file

#### 4 AutoindexHandler

The responsibility of this class is to create an autoindex page
- Receives a directory and builds a autoindex page with a table consisting of file name, size and last modified
- Returns an empty string if error

#### 5 Helper classes

- Directory: encapsulates opening and closing of directory
- FileSystemPolicy: wraps C functions for the file system (mainly stat and directory related). THis makes for easier testing via a MockFileSystemPolicy

#### 6 utils

New utils function:
- getFileExtension - get file extension
- getGMTString - get time string in greenwich meantime
- getLocaltimeString - get time string in localtime
- statusCodeToReasonPhrase - transform status code to reason phrase
- getDefaultErrorPage - construct default error page depending on status code

#### Tests

Included testfiles for AutoindexHandler, ResponseBodyBuilder, TargetResourceBuilder and utils using google mock
Adapted Makefile to compile tests.

### CODE REVIEW

File: test/test_AutoindexHandler.cpp
Comment (QCHR1581) [inline test/test_AutoindexHandler.cpp:1]: Should be deleted. <gmock/gmock.h> is already included

**Review by QCHR1581 — APPROVED** [review]
I like the approach to use a class as a wrapper for c functions. Definitely makes it easier to test and also more coherent. 

Furthermore good amount of tests implemented which showcase the the handlers work as they should.

File: src/ResponseBuilder.cpp
Comment (gwolf-011235) [reply src/ResponseBuilder.cpp:55]: Yes it's a temp solution: The HTTP Request doesn't store the status code right now. When it's added, it will make more sense

File: inc/MockFileSystemPolicy.hpp
Comment (gwolf-011235) [reply inc/MockFileSystemPolicy.hpp:3]: true, added

File: test/test_AutoindexHandler.cpp
Comment (gwolf-011235) [reply test/test_AutoindexHandler.cpp:1]: ture, removed

File: src/ResponseBuilder.cpp
Comment (QCHR1581) [reply src/ResponseBuilder.cpp:55]: Great, thanks for the clarification!

File: test/test_utils.cpp
Comment (u413-284-si) [inline test/test_utils.cpp:9]: Ok, I see now why you check against the dir in the function. The previous this regarding comment can be ignored.

**Review by u413-284-si — CHANGES_REQUESTED** [review]
Generally well structured implementations. Maybe a bit overstructured even.

The response building is straight forward and easily readable. The testing seems also well thought through. Especially the mocking of the system behaviour is very interesting to see in its implementation.

There is a tendency to create many classes and therefore files. Some of those classes I would have incorporated in the ResponseBuilder (e.g. the ResponseBodyHandler or the TargetResourceHandler). But that is up to everyones preference.

Some smaller changes requested, certain parts could be optimised and/or restructured.

File: inc/HTTPResponse.hpp
Comment (gwolf-011235) [reply inc/HTTPResponse.hpp:12]: This is for compatibility with the current version of the request parser, which also has it as a string. 

I could integrate it now, or wait until the request parser has been merged?

File: inc/StatusCode.hpp
Comment (gwolf-011235) [reply inc/StatusCode.hpp:3]: You already checked it out :D

File: src/Directory.cpp
Comment (gwolf-011235) [reply src/Directory.cpp:1]: This stems from my - granted: limited - understanding of object oriented programming.

I tried to follow the concepts Single Responsibility Principle (SRP) and Resource Aquisition Is Initialisation (RAII)

- RAII: technique which binds the life cycle of a resource that must be acquired before use ([...] open file [...]) to the lifetime of an object. (from here: [LINK](https://en.cppreference.com/w/cpp/language/raii)). 

This among other things guarantees that we don't forget to release an accquired resource. I think opening files is the goto example for Stroustrup: you don't open a file in a class directly, you use the class std::filestream. Then you don't have to remember to close the file, because it automatically happens in the destructor. You also know that the file opening failed when the filestream object returns .good() == false or throws (if you enabled it).

In the same vein does the directory class handle the resource acquisition and release. 
By encapsulating the resource acquistion we don't need to handle multiple fail points (resource is not available, resource is not readable, file is not available) and don't need to worry about forgetting to release the resource

- SRP: every class, module, or function in a program should have one responsibility/purpose in a program 

The class Autoindexhandler has the main responsibility of creating an index out of files in a directory. It checks file sizes and file changes and puts them into a html document. To get a list of files in a directory we could give it the additional responsibilty to manage the resource DIR* and extract every file name into a string vector. However with the directory class we don't need to worry about resource acquisition/release and can "just" create the index.

Another definition of SRP is "every class should have only one reason to change", where change is in the sense of rewritten. If there would be another interface to open directories (for example in the newer standard std::filesystem) you could implement this in the directory class, and Autoindex would be none the wiser (becaue it doesn't care how it gets its string vector)

Lastly we can more easliy reuse the directory class, if we want to get a list of files in a directory. If we coupled it into the Autoindex Handler, we would need more modifications to reuse it.

File: src/ResponseBodyHandler.cpp
Comment (gwolf-011235) [reply src/ResponseBodyHandler.cpp:63]: This class is just a stump right now, because it just returns a defaultErrorPage every time. Since I don't know how the interface for the custom error pages would look like (how are they stored), its hard to program this part.

Thats why it has a FIXME comment

File: src/ResponseBuilder.cpp
Comment (gwolf-011235) [reply src/ResponseBuilder.cpp:13]: Always :-*

File: src/ResponseBuilder.cpp
Comment (gwolf-011235) [reply src/ResponseBuilder.cpp:33]: Not quite: the return value is constructed from the const char* ptr which points to the raw string buffer in the stringstream. 

> Doesn't this imply unnecessary conversions (from and to std::string)?

Stringstream uses a string buffer, which can be accessed via .str(). Since its a string we have access to its function c_str().

In contrast to classic c strings, std::string doesn't need a zero terminator. It tracks the size of the string itself. This means a std::string can have "Hello\0World" (a string with a zero terminator inbetween) and correctly return 11 as size, where strlen would return 5.
.c_str() returns (in general) a pointer to the raw buffer of the string, with a null terminator guaranteed at the end. To be more efficient this generally means (since C11 is standard I think) that a std::string buffer has a zero terminator at the end, > you don't have to allocate for a c_str() call. 
c_str() also has to have constant complexity, so regardless of string size it should need the same time. Therefore its more or less guaranteed (even pre C11) that you get a pointer to the underlying buffer with a zero terminator, since malloc and copy would not be constant.

Even though the std::string itself doesn't need a zero terminator, the buffer oftentimes is allocated +1 > mainly for compatibility reasons for a lot of C functions.

This means by passing stringstream.str() or stringstream.str().c_str() you can control which constructor of std::string gets called with no extra cost.

> Why not just clear the stringstream?

If we "clear" the stringstream by setting it to empty string we have two performance implications:
1. we need to copy the empty string
2. we don't know what happens with the allocated buffer

ad 1. When we set the stringstream to empty string we need to construct a string object and copy it into the buffer. I don't exactly know how it works but at least in godbolt it shows more code output for `ss.str("");` instead of `ss.seekp(std::ios::beg);`. Interestingly if we just copy an empty string - `ss.str(std::string))` - the difference is smaller.

See here for godbolt compiler explorer: [LINK](https://godbolt.org/z/PjdqPzx6M)

ad 2. If we never clear the buffer and just reuse the previously used thing we dont need to reallocate. I'm not sure if the buffer gets shrinked to size when you copy an empty buffer into it. But if we never clear it it can't happen. This means that if the ResponseBuilder creates a big message, the buffer stays big.

With seekp we just return to the beginning and add new stuff. Since most stuff gets added as if it was a c string, it also sets a new zero terminator.  Example:
```
stringstream ss;
ss << "Hello World";
ss.seekp(std::ios::beg)
ss << "Short";
// ss.str() is now "Short\0World"
```
When we would now use the std::string constructor and the use for example send() we still would only see "Short" since send() expects a zero terminated string, and it is terminated after "Short". But: we would also copy the never used stuff. I we use the const char* we only copy upto the zero temrinator.

> Why not clear it after sending?

We can get the same message multiple times as long as we don't build a new one. I don't see added value in clearing it gets used once. Granted: we won't use it multiple times, but why limit functionality with no added benefit?

File: src/ResponseBuilder.cpp
Comment (gwolf-011235) [reply src/ResponseBuilder.cpp:53]: m_response is the reused stringstream. I could have also use a HTTPResponse struct as member. But since I'd planned to construct the ResponseBuilder only once, I would need to clear the struct for every new response. This seems more error prone than just creating a new one.

Because of Return Value Optimisation (RVO) this call should cost nothing, since the struct gets initialised with the return value and not copied.

The initHTTPResponse right now is just a converter from Request to Response structs, so maybe it will have more logic, than just copying stuff from and to.

File: src/ResponseBuilder.cpp
Comment (gwolf-011235) [reply src/ResponseBuilder.cpp:77]: See explanation above.

File: src/ResponseBuilder.cpp
Comment (gwolf-011235) [reply src/ResponseBuilder.cpp:117]: Bruce Lee? :0

File: src/ResponseBuilder.cpp
Comment (gwolf-011235) [reply src/ResponseBuilder.cpp:144]: This is a temporary.

MIME (Multipurpose Internet Mail Extensions) types indicate the nature and format of a document, file, or assortment of bytes on the server:
- html pages have text/html
- css have text/css
- images can have image/jpeg or image/png

It stems from emails but the name seems to have stuck. 
For a longer explanation of MIME types see for example this artivle from Mozilla [LINK](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types)

They are standardized in IETF's [RFC 6838](https://datatracker.ietf.org/doc/html/rfc6838) 
But they can often be freely defined on the server. For example nginx has the types directive: [LINK](https://nginx.org/en/docs/http/ngx_http_core_module.html#types). This can be done on server and location level

Since the interface on how we would store the mime types is not defined yet, the response builder just has some standard mime types set. It just fetches them from its internal storage. Later on this would use the found location to fetch the correct type from there.

File: src/TargetResourceHandler.cpp
Comment (gwolf-011235) [reply src/TargetResourceHandler.cpp:44]: In nginx if you don't define a location in a server config you get a default location which matches "/" (i.e. everything)
If you set a location "/test", then you don't have a "/" location.

We could implement this by creating a default location "/" which I can also loop through (so I dont need an extra check: "is there a default location?")

The behaviour you're describing is handled in the next three lines below: if we don't find a matching location we set StatusNotFound. This would create the error response you're describing later via in the ResponseBodyHandler and the appendStatusLine() and appendHeaders().

File: src/utils.cpp
Comment (gwolf-011235) [reply src/utils.cpp:22]: Resolved as below

File: inc/utils.hpp
Comment (gwolf-011235) [reply inc/utils.hpp:9]: Macros are more c style way of doing things. 
They don't give you static type checking, since the Preprocessor just takes the macro value and inserts it everywhere it can find the macro. 
By using const int, the compiler can do a type check beforehand and warn if needed. This might not apply here, but is more of a coding style. For example in Core Guidelines ES.31: [Don’t use macros for constants or “functions”](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es31-dont-use-macros-for-constants-or-functions)

File: test/test_utils.cpp
Comment (gwolf-011235) [reply test/test_utils.cpp:9]: Why not just delete the comment :o

File: inc/HTTPResponse.hpp
Comment (gwolf-011235) [reply inc/HTTPResponse.hpp:13]: Chnaged to `IsAutoindex` (like in ConfigFile.hpp)

File: inc/utils.hpp
Comment (gwolf-011235) [reply inc/utils.hpp:7]: Preemptively changed namespace to webutils

File: src/AutoindexHandler.cpp
Comment (gwolf-011235) [reply src/AutoindexHandler.cpp:44]: fixed

File: src/FileSystemPolicy.cpp
Comment (gwolf-011235) [reply src/FileSystemPolicy.cpp:88]: removed

File: src/FileSystemPolicy.cpp
Comment (gwolf-011235) [reply src/FileSystemPolicy.cpp:105]: removed

File: src/FileSystemPolicy.cpp
Comment (gwolf-011235) [reply src/FileSystemPolicy.cpp:120]: removed

File: src/FileSystemPolicy.cpp
Comment (gwolf-011235) [reply src/FileSystemPolicy.cpp:137]: removed

File: src/ResponseBodyHandler.cpp
Comment (gwolf-011235) [reply src/ResponseBodyHandler.cpp:31]: Changed

File: src/utils.cpp
Comment (gwolf-011235) [reply src/utils.cpp:18]: removed

File: src/utils.cpp
Comment (gwolf-011235) [reply src/utils.cpp:39]: Replaced with const std::string& format

File: src/utils.cpp
Comment (gwolf-011235) [reply src/utils.cpp:53]: const std::string& format

File: inc/HTTPResponse.hpp
Comment (u413-284-si) [reply inc/HTTPResponse.hpp:12]: RequestParser is merged and is using the enum for Method. Can be adapted now.

File: inc/utils.hpp
Comment (u413-284-si) [reply inc/utils.hpp]: Created issue to not forget about it.

File: src/Directory.cpp
Comment (u413-284-si) [reply src/Directory.cpp:1]: Nice explanation, very thorough. Point taken.

File: src/ResponseBodyHandler.cpp
Comment (u413-284-si) [reply src/ResponseBodyHandler.cpp:62]: Yes, this is evident and not the cause of the implicit question:

What do you mean by the "right extension"? And a "data field of HTML structs"?

File: inc/HTTPResponse.hpp
Comment (gwolf-011235) [reply inc/HTTPResponse.hpp]: Adapted

File: src/ResponseBuilder.cpp
Comment (u413-284-si) [reply src/ResponseBuilder.cpp:33]: Oida. How much time did you spend to figure out this

File: src/ResponseBuilder.cpp
Comment (u413-284-si) [reply src/ResponseBuilder.cpp:53]: Would advise to rename the member to m_responseStream to make its identity more clear. And to differentiate to the struct.

File: src/ResponseBuilder.cpp
Comment (u413-284-si) [reply src/ResponseBuilder.cpp:117]: No, Bruce Willis from DieHard

File: src/ResponseBuilder.cpp
Comment (u413-284-si) [reply src/ResponseBuilder.cpp:144]: Yes, I meant how do you imagine this location-specific assignment with this function?

File: src/TargetResourceHandler.cpp
Comment (u413-284-si) [reply src/TargetResourceHandler.cpp:44]: To get on the same page:
We are discussing the situation that the requested target resource location is not found among any of the locations in our server (defined by the config file).

In that case you want to go to a default location like "/" and look for the target resource there? Despite the requested location being another?

If yes, the question begs: is that extra step really necessary? Instead of directly defaulting to creating the "not found" answer.

File: test/test_utils.cpp
Comment (u413-284-si) [reply test/test_utils.cpp]: To have more of an exchange of thoughts!

File: test/test_AutoindexHandler.cpp
Comment (u413-284-si) [inline test/test_AutoindexHandler.cpp:1]: The compiler can't find the gmock header.

File: src/ResponseBodyHandler.cpp
Comment (gwolf-011235) [reply src/ResponseBodyHandler.cpp:62]: Preface: this is just a comment for me so that I don't forget. This is not finished. It is a work in progress.

> What do you mean by the "right extension"? 

ResponseBuilder appends header Content-Type depending on located target resource, by checking the file extension. 

In case of error we don't have a targetResource - we had an error, we can't deliver. So we deliver an error page. If no custom error page is set, default error page is sent. The default error page does not exist as a file, it has no file extension. So I just set targetResource to error.html to get the right extension - ".hmtl".

> And a "data field of HTML structs"?

It says "data field of HTML struct." I could also make a std::string in HTTP Response struct. Then init it to ".html". Then overwrite if necessary.

File: src/ResponseBuilder.cpp
Comment (gwolf-011235) [reply src/ResponseBuilder.cpp:33]: I read the article I posted. Played around in Godbolt to confirm.

File: src/ResponseBuilder.cpp
Comment (gwolf-011235) [reply src/ResponseBuilder.cpp:55]: Renamed

File: src/ResponseBuilder.cpp
Comment (gwolf-011235) [reply src/ResponseBuilder.cpp:123]: Ah! I see it more as Tryhard :p

File: src/ResponseBuilder.cpp
Comment (gwolf-011235) [reply src/ResponseBuilder.cpp:150]: More specific question would help.

The two functions initMIMETypes and getMIMEType have the same @todo, since they are both dependent on further implementation details, i.e. how we store the MIME types.

Right now TargetResourceHandler saves reference of found location in HTTP Response. getMIMEType can fetch the map, vector, whatever of this location and search for type.

I imagine two possible implementations (right now):
1. we save in each location all MIME types (specific and default ones)
2. we only save MIME types we have explicitly defined and have defaults in ResponseBuilder

In case 1 we don't need this map, so I would delete. In case 2 initMIMETypes inits the map of types saved in ResponseBuilder. Maybe there are other possibilities.

File: src/TargetResourceHandler.cpp
Comment (gwolf-011235) [reply src/TargetResourceHandler.cpp:44]: > We are discussing the situation that the requested target resource location is not found among any of the locations in our server (defined by the config file).

Yes.

> In that case you want to go to a default location like "/" and look for the target resource there? Despite the requested location being another?

Its more of an implementation detail: Say you have only one location block "/secret" and you request "/something". It would match nothing. 

We could then just report Not found and leave (this is currently implemented). This means we have to always explicitly define all locations.

We could also have a default location "/" which is always implicitly there. So the ConfigFileParser would just init this location. Since a HTTP request has to have a slash, this location would be the catch-all for all requests.
We would still search for /something but in the default location. 

Consider nginx: when you have a very plain config:
```
http {
  server {
    root /data/www;
  }
} 
```

nginx will serve you, when you request "/" --> it will give you /data/www/index.html. Even though we didn't specify any location.
Even if you do specify one location like
```
http {
  server {
    root /data/www;
    location /secret {
    }
  }
} 
```
nginx will still serve you, when you request "/" --> it will give you /data/www/index.html.
It will also serve you, when you request "/something" --> it will give you /data/www/something/index.html

I assume the location for "/" is always defined implicitly. (also: index.html seems to be also implicitly defined )

What doesn't work for me is:
```
http {
  server {
  }
} 
```
-> no root specified. But it still is a valid config. I suspect that the default root, which should be "html" was changed and not updated in manual.

File: test/test_AutoindexHandler.cpp
Comment (gwolf-011235) [reply test/test_AutoindexHandler.cpp:1]: You need to rebuild your devcontainer

File: src/ResponseBuilder.cpp
Comment (u413-284-si) [reply src/ResponseBuilder.cpp:33]: Which article? There is only the link to Godbolt.

File: src/ResponseBuilder.cpp
Comment (u413-284-si) [reply src/ResponseBuilder.cpp:150]: Best to talk about this in person to clear it up.

File: src/TargetResourceHandler.cpp
Comment (u413-284-si) [reply src/TargetResourceHandler.cpp:44]: > It will also serve you, when you request "/something" --> it will give you /data/www/something/index.html

Even if we don't have a folder "something" inside of /data/www/ ?

File: src/ResponseBuilder.cpp
Comment (gwolf-011235) [reply src/ResponseBuilder.cpp:33]: This one: http://cplusplus.bordoon.com/speeding_up_string_conversions.html 

Posted in Discord regarding reusing stringstream.
The article mentions using seekp and return as c_str().
It does not spell everything out, which is why I checked with godbolt > for example it doesn't say anything about "disadvantage" of setting it to "".

File: src/TargetResourceHandler.cpp
Comment (gwolf-011235) [reply src/TargetResourceHandler.cpp:44]: The target would be /data/www/something/index.html. 
It gets checked in the following try block at line 56 ff
If there is no such file then it will 404 File not found.

If we don't want to go with the default "/" location we could exit early (this is impl right now)
If we implement the default location we would of course need to check if the file exists (in the following lines)

File: src/ResponseBodyHandler.cpp
Comment (u413-284-si) [reply src/ResponseBodyHandler.cpp:62]: Discussed and cleared up.

File: src/TargetResourceHandler.cpp
Comment (u413-284-si) [reply src/TargetResourceHandler.cpp:44]: After discussion agreed on integration of default location.

**Review by u413-284-si — APPROVED** [review]
Agreed to resolve specific conversation topics in future branches, documented in respective issues.

**Pull request merged**
